### PR TITLE
#164139782 Pagination Support for Recipes

### PR DESCRIPTION
**Pull request by LordUche**
#### What does this PR do?
- adds pagination support for listing recipes
#### Description of Task to be completed?
- write unit tests
- accept `page` and `limit` query parameters from request url
- validate query parameters with `Joi`
- document feature with swagger
#### How should this be manually tested?
- ensure build is passing on Travis
- send a *GET* request to `/api/recipes` and pass in `page` and/or `limit` query params
#### Any background context you want to provide?
Pagination support is lacking for our recipes. Reading can be a nightmare when we have so many recipes.
#### What are the relevant pivotal tracker stories?
[#164139782](https://www.pivotaltracker.com/story/show/164139782)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

### CODE REVIEW

**Review by andela-pessien — COMMENTED** [review]
I think I have expressed my issue with including and pushing changes from unmerged feature branches in one's own feature branch more than often enough.

As it stands I cannot approve this PR because it contains changes I have commented on in a different PR and unless corrections based on the feedback given reflect here, it cannot get merged. I can't even review it properly because far too much of it is not your own work.

**Comment by andela-pessien** [comment]
@LordUche Please update this - the corresponding PR has been merged

**Comment by LordUche** [comment]
@andela-pessien I needed to have the `list recipes` functionality working for pagination to work. That was why I branched out of that feature branch. I have rebased with `develop` successfully and dropped all commits from the feature branch now.

File: server/controllers/recipe-controller.js [hunk 1]
```diff
@@ -148,20 +152,30 @@ class RecipeController {
    * @param {Object} res Express response object
    * @returns {undefined}
    */
-  static getRecipes(req, res) {
-    Recipe.findAll(defaultRecipeDbFilter)
+  static async getRecipes(req, res) {
+    const { page } = req.query;
```
Comment (andela-pessien) [inline server/controllers/recipe-controller.js]: This doesn't match the spec's requirements - this endpoint should accept an _offset_, not a _page_.

File: test/recipes.spec.js [hunk 2]
```diff
@@ -141,4 +158,156 @@ describe('Recipes', () => {
...
+    });
+  });
+
+  describe('Pagination', () => {
+    const url = '/api/recipes';
+    let items;
+    before(async () => {
+      const array = Array(50)
+        .fill()
+        .map(() => {
```
[the remark this answers is not in the document]
Comment (andela-pessien) [reply test/recipes.spec.js]: Please address this

File: test/recipes.spec.js [hunk 3]
```diff
@@ -333,4 +331,115 @@ describe('Recipes', () => {
...
+    it('should get the second set of 10 recipes when page is 2 and no limit is set', done => {
+      chai
+        .request(server)
+        .get(`${url}?page=2`)
+        .end((err, res) => {
+          const { recipes } = res.body;
+          expect(res).to.have.status(200);
+          expect(recipes).to.be.an('Array');
+          expect(recipes.length).to.equal(10);
+          expect(recipes[0].id).to.not.equal(items[0].id);
```
Comment (andela-pessien) [inline test/recipes.spec.js]: You should test positives, not just negatives. (i.e. the fact that the first returned recipe is not the first created recipe does not guarantee whether the right recipes were returned or not). That applies to the rest of the examples after this here as well

File: test/recipes.spec.js [hunk 4]
```diff
@@ -333,4 +331,115 @@ describe('Recipes', () => {
...
+      await Promise.all(array);
+    });
+
+    it('should get all recipes when no page or limit is set', done => {
+      chai
+        .request(server)
+        .get(url)
+        .end((err, res) => {
+          const { recipes } = res.body;
+          items = recipes;
```
Comment (andela-pessien) [inline test/recipes.spec.js]: You shouldn't be performing setup for other examples inside a test example, as the examples in a suite should be runnable in _any_ order (or even in isolation). That's what the pre- and post-hooks are for.

**Review by andela-pessien — CHANGES_REQUESTED** [review]

**Review by andela-pessien — APPROVED** [review]

**Pull request merged**